File: articles/calculus-for-systems-modeling/articles/differential-equations-and-dynamic-systems/c/dynamic_system_audit.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

double exponential_rate(double x, double r){ return r*x; }
double logistic_rate(double x, double r, double k){ return r*x*(1.0 - x/k); }

void simulate(const char* scenario, int logistic){
  double x = 10.0;
  double r = 0.35;
  double k = 100.0;
  double dt = 0.1;
  int steps = 100;
  for(int n=0; n<=steps; n++){
    double t = n*dt;
    double rate = logistic ? logistic_rate(x,r,k) : exponential_rate(x,r);
    printf("%s,%s,%.6f,%.6f,%.6f,%.6f,%.6f,%s,%s\n",
      scenario,
      logistic ? "dx_dt_equals_r_x_one_minus_x_over_K" : "dx_dt_equals_r_x",
      t, x, rate, r, logistic ? k : -1.0, "explicit_euler",
      logistic ? "Logistic growth assumes a fixed carrying capacity." : "Exponential growth assumes no capacity constraint.");
    x = x + dt*rate;
  }
}

int main(void){
  printf("scenario,model_type,time,state,rate,growth_rate,carrying_capacity,method,warning\n");
  simulate("exponential_growth", 0);
  simulate("logistic_growth", 1);
  return EXIT_SUCCESS;
}

File: articles/calculus-for-systems-modeling/articles/differential-equations-and-dynamic-systems/fortran/dynamic_system_audit.f90
program dynamic_system_audit
  implicit none
  call simulate('exponential_growth', .false.)
  call simulate('logistic_growth', .true.)
contains
  real(8) function exp_rate(x,r)
    real(8), intent(in) :: x,r
    exp_rate = r*x
  end function
  real(8) function log_rate(x,r,k)
    real(8), intent(in) :: x,r,k
    log_rate = r*x*(1.0d0 - x/k)
  end function
  subroutine simulate(label, logistic)
    character(len=*), intent(in) :: label
    logical, intent(in) :: logistic
    real(8) :: x, r, k, dt, t, rate
    integer :: n, steps
    x = 10.0d0; r = 0.35d0; k = 100.0d0; dt = 0.1d0; steps = 100
    do n = 0, steps
      t = n*dt
      if (logistic) then
        rate = log_rate(x,r,k)
        print '(A,1X,A,5F12.6,1X,A)', trim(label), 'logistic', t, x, rate, r, k, 'explicit_euler'
      else
        rate = exp_rate(x,r)
        print '(A,1X,A,5F12.6,1X,A)', trim(label), 'exponential', t, x, rate, r, -1.0d0, 'explicit_euler'
      end if
      x = x + dt*rate
    end do
  end subroutine
end program
